### PR TITLE
feat(frontend): fuzzy, separator-insensitive search across pickers

### PR DESCRIPTION
**Pull request by kristofferremback**
## Problem

Every client-side picker matches on strict substrings. `scoreMatch` (`apps/frontend/src/lib/match-score.ts`) and `scoreStreamMatch` (`lib/stream-sort.ts`) tier exact/prefix/contains over `toLowerCase()` — so the emoji picker misses `thumbs up` and `thumbsup` for `thumbs_up`, the quick switcher misses `growthplan` for `#growth-plan`, and abbreviation queries (`grpl`, `thup`) find nothing. Seven more surfaces (members tab, label picker, board filter pickers, search filter menu, quick-switcher filter selects, target-picker new-options, virtual DM rows) hand-rolled the same boolean `.includes()` inline.

## Solution

One new primitive module, `lib/fuzzy.ts`, slotted into the two shared scorers as lower-priority tiers — fuzzy can only *append* results, never displace a substring match — plus a sweep of the inline `.includes()` stragglers onto the shared scorer.

### How it works

1. **`foldSeparators(text)`** strips `[\s_-]+`, making whitespace/dash/underscore interchangeable: `thumbs up` ≡ `thumbs-up` ≡ `thumbsup` ≡ `thumbs_up`.
2. **`fuzzyQuality(query, text)`** scores an in-order subsequence match via a small DP (O(len(query)·len(text))): each matched char scores 1, +1 when it extends a contiguous run or sits at a word boundary; quality = best total / (2·query length), in (0, 1]. `thup` → `thumbs_up` scores 1.0 (t-h contiguous from start, u-p contiguous from the `_` boundary); `inv` scattered across "Open a side-conversation" scores 0.67.
3. **`scoreMatch` tier ladder** (lower = better): label exact(0) < norm-exact(1) < prefix(2) < norm-prefix(3) < contains(4) < norm-contains(5); keywords the same ladder one band below (6–11); label fuzzy 12+(1−quality); keyword fuzzy 13+(1−quality). All previous relative orderings are preserved — the old 0/1/2 tiers map onto 0/2/4.
4. **`FUZZY_MIN_QUALITY = 0.7`** rejects scatter-noise. Found in self-review via a failing test: `inv` fuzzy-matched the *description* "Open a side-conversation" and polluted the slash-command menu. Compact abbreviation matches land near 1.0; thin scatter across sentences lands well under 0.7.
5. **`scoreStreamMatch` delegates to `scoreMatch`**, keeping the raw stream-id substring as a last-resort tier (score 100, below the fuzzy band — previously it sat directly after contains).
6. Everything already routed through `rankMatches`/`scoreStreamMatch` upgrades for free: emoji picker, quick switcher (streams + commands), mention/`#channel`/slash-command/arg-picker triggers, `in:#` filter, share modal, share picker, explorer stream filter, composer target picker.

### Key design decisions

**1. Hand-rolled scorer over a dependency.** No fuzzy library exists in the repo (fuse.js/fzf/match-sorter all absent); the need is ~60 lines of separator folding + subsequence DP with full unit coverage. A library brings its own scoring model that would fight the existing tier ladder the surfaces already rely on (label-above-keyword, exact-above-prefix).

**2. Fuzzy as a strictly-worse band, not a replacement scorer.** Rewriting the tiers around a single fuzzy score would reshuffle results users already know. Appending fuzzy below every substring tier means existing queries return identical top results; only previously-empty tails gain entries.

**3. Quality floor at 0.7 instead of per-surface gating.** The alternative — fuzzy only on labels, not keywords — would lose real matches (emoji alias abbreviations) to kill noise the floor already kills. The floor is centralized in `bestFuzzy`, so every surface inherits the same junk cutoff.

**4. Straggler sweep to the shared scorer (INV-35).** The seven inline `.includes()` sites now call `rankMatches`/`scoreMatch`. Board filter pickers keep their own selected-first sort and use `scoreMatch(...) !== Infinity` as a boolean, since reordering by score would fight the selected-pinning.

## New files

| File | Purpose |
| ---- | ------- |
| `apps/frontend/src/lib/fuzzy.ts` | `foldSeparators` + `fuzzyQuality` subsequence scorer |
| `apps/frontend/src/lib/fuzzy.test.ts` | Unit coverage: folding, boundary bonuses, scatter vs compact, edge cases |

## Modified files

| File | Change |
| ---- | ------ |
| `lib/match-score.ts` | Tier ladder grew normalized + fuzzy bands; `FUZZY_MIN_QUALITY` floor |
| `lib/match-score.test.ts` | Updated tier numbers; new normalization/fuzzy/ordering cases |
| `lib/stream-sort.ts` | `scoreStreamMatch` delegates to `scoreMatch`; id fallback → 100 |
| `lib/stream-sort.test.ts` | Ordering-based assertions + separator cases |
| `lib/emoji-picker.test.ts` | Separator + fuzzy regression tests for `filterBySearch` |
| `components/search/search-filter-menu.tsx` | ChannelPicker inline filter → `rankMatches` over `[slug, name]` |
| `components/quick-switcher/filter-select.tsx` | UserSelect/StreamSelect → `rankMatches` |
| `components/quick-switcher/use-stream-items.tsx` | Virtual DM rows → `scoreMatch` |
| `components/stream-settings/members-tab.tsx` | Member search → `rankMatches` over `[name, slug]` |
| `components/board/board-filter-pickers.tsx` | Stream/label scope pickers → `scoreMatch` boolean (keeps selected-first sort) |
| `components/labels/label-picker.tsx` | Label filter → `rankMatches` |
| `components/composer/stream-target-picker.tsx` | NEW_OPTIONS filter → `scoreMatch` |
| `components/editor/triggers/use-command-arg-picker.test.ts` | `opus` now legitimately fuzzy-admits `anthropic/claude-sonnet-4` below the substring match; assert ranking instead of exact set |

## Out of scope (deliberate)

- `SearchableSelect` (timezone picker, add-member/add-bot pickers) already gets fuzzy-ish matching from cmdk's built-in `command-score`; left as is.
- Server-side search (messages, memos, attachments) — different layer, ILIKE/similarity on the backend.
- No typo tolerance (edit distance). Subsequence + separator folding covers the reported misses; Levenshtein is a separate, noisier tradeoff.

## Test plan

- [x] `bun run test` (frontend) — 4403 pass, 0 fail
- [x] Full monorepo typecheck + eslint via pre-commit hook — clean
- [x] Live browser verification (dev:test stack + Playwright): emoji picker finds 👍 for `thumbs up` / `thumbs-up` / `thumbsup` / `thup`, 🙄 for `eye roll` / `rolleyes`; junk query `xqzv` still shows the empty state. Quick switcher finds `#growth-plan` for `growth plan` / `growthplan` / `grpl` and `#project-alpha` for `pralpha`.
- [ ] Straggler surfaces (members tab, board pickers, label picker) verified via unit tiers + typecheck only — not driven individually in the browser; they call the same scorer the driven surfaces do.

---

🤖 _PR by [Claude Code](https://claude.com/claude-code)_

https://claude.ai/code/session_01NHpwiZxCcVBfTVQLVtiYUw

---
<a href="https://app.blacksmith.sh/threahq/codesmith/threa/pr/1355"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1786773297&installation_id=129946197&pr_number=1355&repository=threahq%2Fthrea&return_to=https%3A%2F%2Fgithub.com%2Fthreahq%2Fthrea%2Fpull%2F1355&signature=eddf690b72d6da705b083e85c6ed6055624b464583aec248b19ca357c30c0152"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: apps/frontend/src/components/stream-settings/members-tab.tsx [hunk 1]
```diff
@@ -79,11 +80,10 @@ export function MembersTab({ workspaceId, streamId, currentUserId }: MembersTabP
...
 
-  const filteredMembers = useMemo(() => {
-    if (!search) return enrichedMembers
-    const q = search.toLowerCase()
-    return enrichedMembers.filter((m) => m.name.toLowerCase().includes(q) || m.slug.toLowerCase().includes(q))
-  }, [enrichedMembers, search])
+  const filteredMembers = useMemo(
+    () => rankMatches(enrichedMembers, search, (m) => ({ labels: [m.name, m.slug] })),
+    [enrichedMembers, search]
+  )
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply apps/frontend/src/components/stream-settings/members-tab.tsx:86]: Fixed in e00a9f7a, but centrally rather than at this call site: `scoreMatch` and `rankMatches` now trim the query before scoring, so a whitespace-only query behaves as empty for every consumer (this tab, the quick-switcher filter selects, the search-filter-menu channel picker, etc.) instead of each site remembering to `.trim()` (INV-35). Covered by new cases in `match-score.test.ts` ("returns 0 for an empty or whitespace-only query", "returns the input unchanged for an empty or whitespace-only query").

🤖 _Response by [Claude Code](https://claude.com/claude-code)_

File: apps/frontend/src/lib/match-score.test.ts [hunk 2]
```diff
@@ -8,33 +8,74 @@ describe("scoreMatch", () => {
...
+  })
+
+  it("ranks label fuzzy above keyword fuzzy, and compact fuzzy above scattered", () => {
+    const labelFuzzy = scoreMatch("thup", ["thumbs_up"])
+    const keywordFuzzy = scoreMatch("thup", ["Something Else"], ["thumbs_up"])
+    expect(labelFuzzy).toBeLessThan(keywordFuzzy)
+
+    const compact = scoreMatch("cat", ["cat_face"])
+    const scattered = scoreMatch("cat", ["congratulations"])
+    expect(compact).toBeLessThan(scattered)
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply apps/frontend/src/lib/match-score.test.ts:64]: Fixed in e00a9f7a. `match-score.test.ts`: the label-vs-keyword-fuzzy and compact-vs-scattered comparisons now assert the right-hand side is an admitted match (`not.toBe(Infinity)`) first — previously a broken fuzzy band returning Infinity on the right would have passed the bare `lessThan`. `emoji-picker.test.ts`: the fuzzy test now includes a raw-substring candidate (`setup`) and asserts the full ordered array `["setup", "thumbs_up", "tulip"]` — substring tier first, then fuzzy by quality — instead of a Set.

🤖 _Response by [Claude Code](https://claude.com/claude-code)_

File: apps/frontend/src/lib/stream-sort.test.ts [hunk 3]
```diff
@@ -58,16 +58,29 @@ describe("scoreStreamMatch", () => {
...
-    expect(scoreStreamMatch(general, "#general")).toBe(0)
-    expect(scoreStreamMatch(generic, "#gen")).toBe(1)
-    expect(scoreStreamMatch(ageneral, "general")).toBe(2)
-    expect(scoreStreamMatch(idMatch, "general")).toBe(3)
+    const scores = [
+      scoreStreamMatch(general, "#general"),
+      scoreStreamMatch(generic, "#gen"),
+      scoreStreamMatch(ageneral, "general"),
+      scoreStreamMatch(fuzzy, "grpl"),
+      scoreStreamMatch(idMatch, "general"),
```
[the remark this answers is not in the document]
Comment (kristofferremback) [reply apps/frontend/src/lib/stream-sort.test.ts:73]: Half holds, half fixed in e00a9f7a. `use-command-arg-picker.test.ts`: real gap — added `expect(values).toContain("anthropic/claude-sonnet-4")` so removing the fuzzy tail now fails the test. `stream-sort.test.ts`: the fuzzy entry's presence is already asserted — `scores.every((s) => s !== Infinity)` includes the `grpl → growth-plan` score, so removing fuzzy matching turns that score to Infinity and fails the assertion; left as is.

🤖 _Response by [Claude Code](https://claude.com/claude-code)_
